### PR TITLE
Electron app packaging

### PR DESCRIPTION
**Pull request by notdanilo**
### Pull Request Description
This PR provides Electron app packaging and unified an unified framework for web development and Electron app development.

It also merges works from `wip/mwu/fm-ci`, which provides Electron app and File Manager Mock Server executables as Github Action artifacts.

### Important Notes
* The file structure was reorganized.
* HTML & Javascript code was slightly reworked to be reused both in web and Electron app envs

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: app/.yarnclean [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# test directories
```
Comment (wdanilo) [inline app/.yarnclean]: Why are we using yarn instead of npm here? Do we have to migrate to yarn? I mean, I dont want to use npm and yarn side by side as they do mostly the same thing.

File: app/package.json [hunk 2]
```diff
@@ -11,7 +10,14 @@
...
     "start-build-time-profile": "webpack-dev-server --mode development --profile --progress",
-    "prod-server": "npx http-server -p 3000 --proxy http://localhost:3000\\? dist"
+    "prod-server": "npx http-server -p 3000 --proxy http://localhost:3000\\? dist",
+    "dev": "electron-webpack dev",
+    "compile": "electron-webpack",
+    "dist": "yarn compile && electron-builder",
+    "dist:dir": "yarn dist --dir -c.compression=store -c.mac.identity=null"
+  },
+  "dependencies": {
+    "source-map-support": "^0.5.12"
```
Comment (wdanilo) [inline app/package.json]: do we need this in deps? I'm trying to keep deps as small as possible (zero) in order to keep binary small.

File: app/package.json [hunk 3]
```diff
@@ -11,7 +10,14 @@
     "build-dev": "webpack --mode development",
     "start": "webpack-dev-server --mode development",
     "start-build-time-profile": "webpack-dev-server --mode development --profile --progress",
-    "prod-server": "npx http-server -p 3000 --proxy http://localhost:3000\\? dist"
+    "prod-server": "npx http-server -p 3000 --proxy http://localhost:3000\\? dist",
+    "dev": "electron-webpack dev",
+    "compile": "electron-webpack",
+    "dist": "yarn compile && electron-builder",
+    "dist:dir": "yarn dist --dir -c.compression=store -c.mac.identity=null"
```
Comment (wdanilo) [inline app/package.json]: we need better names for that. It's not clear now which commands output web and which electron. I'd use `electron:dev` instead of `dev`. We can do the same for web, like `web:start`

File: app/src/main/index.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+'use strict'
+
+import { app, BrowserWindow } from 'electron'
+import * as path from 'path'
+import { format as formatUrl } from 'url'
+
+const isDevelopment = process.env.NODE_ENV !== 'production'
+
+// global reference to mainWindow (necessary to prevent window from being garbage collected)
+let mainWindow
```
Comment (wdanilo) [inline app/src/main/index.js]: Instead of comment use var like `main_window_keep_alive`.

File: app/src/main/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+function createMainWindow() {
```
Comment (wdanilo) [inline app/src/main/index.js]: btw I know that JS is cammel-case lang, but in other js files we use rust naming convention. What we do about it?

File: app/src/main/index.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  const window = new BrowserWindow({webPreferences: {nodeIntegration: true}})
+
+  if (isDevelopment) {
+    window.webContents.openDevTools()
+  }
+
+  if (isDevelopment) {
+    window.loadURL(`http://localhost:${process.env.ELECTRON_WEBPACK_WDS_PORT}`)
+  }
+  else {
```
Comment (wdanilo) [inline app/src/main/index.js]: - nl

File: app/src/main/index.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+const isDevelopment = process.env.NODE_ENV !== 'production'
+
+// global reference to mainWindow (necessary to prevent window from being garbage collected)
+let mainWindow
+
+function createMainWindow() {
+  const window = new BrowserWindow({webPreferences: {nodeIntegration: true}})
+
+  if (isDevelopment) {
+    window.webContents.openDevTools()
```
Comment (wdanilo) [inline app/src/main/index.js]: in other js files we use 4 space indent

File: app/src/renderer/index.js [hunk 8]
```diff
@@ -0,0 +1,2 @@
+// Initial welcome page. Delete the following line to remove it.
```
Comment (wdanilo) [inline app/src/renderer/index.js]: Are u sure this file should be committed?

File: app/yarn.lock [hunk 9]
```diff
@@ -0,0 +1,7975 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (wdanilo) [inline app/yarn.lock]: Either we use yarn or npm. Having 2 lock files is crazy. I vote for npm as yarn was a thing in npm's early days. Nowadays npm can do everything and is the default tool in this ecosystem.

File: app/.yarnclean (hunk 1 above)
Comment (notdanilo) [reply app/.yarnclean]: @wdanilo it came with Electron's quickstart repo. I am cleaning everything up now

File: app/package.json (hunk 2 above)
Comment (notdanilo) [reply app/package.json]: @wdanilo it seems to be part of Electron's loader. I couldn't workaround it. More research is needed.

File: app/src/main/index.js (hunk 5 above)
Comment (notdanilo) [reply app/src/main/index.js]: I fixed it to snake-case

File: app/src/renderer/index.js (hunk 8 above)
Comment (notdanilo) [reply app/src/renderer/index.js]: It was removed

File: app/src/main/index.js (hunk 7 above)
Comment (notdanilo) [reply app/src/main/index.js]: I fixed the style of this file.

File: app/yarn.lock (hunk 9 above)
Comment (notdanilo) [reply app/yarn.lock]: yarn related files and code were removed.

File: app/package.json (hunk 2 above)
Comment (wdanilo) [reply app/package.json]: Can you provide some sources for that? Where do you know its part of electron loader and what is electron loader exactly?

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -119,3 +120,119 @@ jobs:
...
+          override: true
+      - name: Install wasm-pack
+        uses: actions-rs/cargo@v1
+        with:
+          command: install
+          args: wasm-pack    
+      - name: Set up Python
+        uses: actions/setup-python@v1
+        with:
+          python-version: 3.8
```
Comment (wdanilo) [inline .github/workflows/build.yml]: some steps have no nl, some have single nl, some have double nl here. It's not consistent.

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -119,3 +120,119 @@ jobs:
...
+        name: Upload Artifacts (Ubuntu, AppImage)
+        with:
+          name: Enso IDE (Linux, AppImage)
+          path: app/dist/enso-ide-0.1.0.AppImage
+        if: matrix.os == 'ubuntu-latest'
+
+      - uses: actions/upload-artifact@v1
+        name: Upload Artifacts (Ubuntu, unpacked)
+        with:
+          name: Enso IDE (Linux, unpacked)
```
Comment (wdanilo) [inline .github/workflows/build.yml]: If I understand correctly, "unpacked" means that it was packed in the past and is not packed anymore (https://dictionary.cambridge.org/dictionary/english/unpacked). If I'm right, this is incorrect usage of the word. The correct would be "not packed".

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -119,3 +120,119 @@ jobs:
...
+      - uses: actions/upload-artifact@v1
+        name: Upload Artifacts (macOS, app)
+        if: matrix.os == 'macos-latest'
+        with:
+          name: Enso IDE (macOS, app)
+          path: app/dist/mac/enso-ide.app
+
+
+  package-fm:
+    name: Package File Manager
```
Comment (wdanilo) [inline .github/workflows/build.yml]: What is file manager and why we are creating a package for it?

File: README.md [hunk 13]
```diff
@@ -3,16 +3,17 @@
 [![Coverage](https://img.shields.io/codecov/c/github/luna/basegl?label=Coverage&labelColor=2c3239)](https://codecov.io/gh/luna/basegl/branch/master) 
 ![Stability](https://img.shields.io/static/v1?label=Stability&message=Unstable&color=d52229&labelColor=2c3239)
 
-# BaseGL
+# Enso Studio
 
-BaseGL is a blazing fast 2D vector rendering engine with a rich set of
+Enso Studio is an IDE for hybrid visual and textual functional programming.
+
+Its code uses BaseGL, a blazing fast 2D vector rendering engine with a rich set of
```
Comment (wdanilo) [inline README.md]: I believe this is not the best way of changing it for several reasons:

- The fact that Enso Studio uses Basegl is not important to anyone reading this documentation. Sure we can mention it, but then writing a lot about BaseGL does not make sense. Telling later "This repository is a work in progress. Please refer to BaseGL 1.0 repository for more information:" makes even less sense in such context.
- The documentation should be crisp and concise.

I understand why you want this change, but it should be done different way around:

1. All description of Basegl should be moved to `lib/basegl/README.md`. Later we should refactor different libraries to `lib/basegl`, just as we have `lib/ide`.
2. The top-level `README.md` should describe only Enso Studio and how we should build it (like it describes it now). It can mention in 1 sentence that the GUI bases on BaseGL with a link to `lib/basegl.README.md`.

Let's always think about all the libs as separate things which we will refactor in the future to separate repos.

File: README.md [hunk 14]
```diff
@@ -36,27 +37,61 @@ In particular, you can provide them with `--release`, `--dev`, or `--profile`
...
+```bash
+ide/app$ npm install
+ide/app$ npm run electron:dev
+```
+
+To debug a subsystem, use the `--debug` argument.
+
+```bash
+ide/app$ npm run electron:dev -- --debug        # Shows the selection screen.
+ide/app$ npm run electron:dev -- --debug=shapes # Runs shapes' subsystem.
```
Comment (wdanilo) [inline README.md]: what does "subsystem" word mean in this context?

File: README.md [hunk 15]
```diff
@@ -36,27 +37,61 @@ In particular, you can provide them with `--release`, `--dev`, or `--profile`
...
+```
+
+If you want to distribute the desktop application, run the following steps instead:
+
+```bash
+ide/app$ npm install
+ide/app$ npm run electron:dist
+```
+
+The generated executable will be available at `dist/enso-ide-major.minor.*`. Its extension will
```
Comment (wdanilo) [inline README.md]: to follow unix convention, all variables in names should bre capitalized: `dist/enso-ide-MAJOR.MINOR.*`.

File: README.md [hunk 16]
```diff
@@ -36,27 +37,61 @@ In particular, you can provide them with `--release`, `--dev`, or `--profile`
 flags to switch the compilation profile. If not option is provided, the scripts
 default to the `--release` profile.
 
-### Running application and examples
 For best experience, it is recommended to use the 
-`scripts/watch.sh --dev` in a second shell. In order to build the IDE application, 
-follow the steps below:
+`scripts/watch.sh --dev` in a second shell.
+
+### Web Application
```
Comment (wdanilo) [inline README.md]: As we get this more complex than before, we should make this doc chronological

1. Building Rust Sources.
2. Minimizing the WASM binary size (optional).
3. Running / Packaging the App.

Lets use these 3 titles to group the steps, ok? :)

:

File: app/package.json [hunk 17]
```diff
@@ -1,17 +1,22 @@
...
   "description": "Example Scenes",
-  "main": "index.js",
   "bin": {
     "create-wasm-app": ".bin/create-wasm-app.js"
   },
   "scripts": {
     "build": "webpack --mode production",
     "build-dev": "webpack --mode development",
-    "start": "webpack-dev-server --mode development",
+    "web:dev": "npm run build-dev && webpack-dev-server --mode development",
```
Comment (wdanilo) [inline app/package.json]: Why this change? `+npm run build-dev && `?

File: app/package.json [hunk 18]
```diff
@@ -1,17 +1,22 @@
...
     "start-build-time-profile": "webpack-dev-server --mode development --profile --progress",
-    "prod-server": "npx http-server -p 3000 --proxy http://localhost:3000\\? dist"
+    "web:prod": "npx http-server -p 3000 --proxy http://localhost:3000\\? static/main",
+    "electron:dev": "npm run build-dev && electron-webpack dev",
+    "electron:dist": "npm run build && electron-webpack && electron-builder",
+    "electron:dist:dir": "npm run dist --dir -c.compression=store -c.mac.identity=null"
```
Comment (wdanilo) [inline app/package.json]: This command has a non-obvious name. `dist` is obvious, but `dist:dir` is not

File: app/package.json [hunk 19]
```diff
@@ -1,17 +1,22 @@
 {
-  "name": "example-scenes",
+  "name": "enso-ide",
   "version": "0.1.0",
   "description": "Example Scenes",
-  "main": "index.js",
```
Comment (wdanilo) [inline app/package.json]: Why this change? As far as I know, Webpack config should have `main` defined.

File: app/src/index.html [hunk 20]
```diff
@@ -12,5 +12,5 @@
     <body style="background-color:#DDD">
         <noscript>This page requires JavaScript to run. Please enable it in your browser.</noscript>
     </body>
-    <script type="module" src="/assets/index.js"></script>
```
Comment (wdanilo) [inline app/src/index.html]: Why `-type="module"`?
I like `-/assets`

File: app/src/index.js [hunk 21]
```diff
@@ -2,11 +2,24 @@
 /// user with a visual representation of this process (welcome screen). It also implements a view
 /// allowing to choose a debug rendering test from.
 
+var is_electron = window && window.process && window.process.type;
+var static_path;
+
+if (is_electron) {
+    // To avoid compile-time evaluation
+    eval("window.NODE_ENV=process.env.NODE_ENV; window.DIR_NAME = __dirname");
```
Comment (wdanilo) [inline app/src/index.js]: who will compile-time evaluate it? This sounds like a terrible hack tbh. Are there any docs telling to do it this way?

File: app/src/index.js [hunk 22]
```diff
@@ -2,11 +2,24 @@
...
+    if (NODE_ENV == "development") {
+        static_path = "./"
+    } else {
+        static_path = DIR_NAME.replace(/app\.asar$/, 'static')+"/";
+    }
```
Comment (wdanilo) [inline app/src/index.js]: How does it work? What happens when you replace `app.asar` with `static` and why we are doing it?

File: app/src/index.js [hunk 23]
```diff
@@ -2,11 +2,24 @@
...
+} else {
+    static_path = "./"
+}
+
 import * as loader_module from './loader'
 import * as html_utils    from './html_utils'
```
Comment (wdanilo) [inline app/src/index.js]: Please, never use imports in the middle of a file.

File: app/src/index.js [hunk 24]
```diff
@@ -125,8 +138,7 @@ function show_debug_screen(wasm,msg) {
 
 /// Main entry point. Loads WASM, initializes it, chooses the scene to run.
 async function main() {
-    let target = window.location.href.split('/')
-    target.splice(0,3)
+    let target = window.location.search.substr(1).split("=");
```
Comment (wdanilo) [inline app/src/index.js]: This part `window.location.search.substr(1)` should be refactored to a variable telling what it is.

File: app/webpack.config.js [hunk 25]
```diff
@@ -10,7 +10,7 @@ module.exports = {
         wasm_imports: './src/wasm_imports.js',
     },
     output: {
-        path: path.resolve(root,'dist','assets'),
+        path: path.resolve(root,'static','main'),
```
Comment (wdanilo) [inline app/webpack.config.js]: Why this change? The generated things should not go to a folder named `static`. It is a widely used convention it goes to `dist`. Moreover, `static/main` is something I would never thought will contain a generated set of JS files.

File: app/webpack.config.js [hunk 26]
```diff
@@ -26,10 +26,9 @@ module.exports = {
         ]),
     ],
     devServer: {
-        publicPath: '/assets/',
-        historyApiFallback: {
-            index: '/assets/'
-        }
+        contentBase: './static/main',
+        hot: true,
+
```
Comment (wdanilo) [inline app/webpack.config.js]: Why this change? I'm asking **about each changed line here**. Maybe it's OK, but I feel it's not and I want to understand it.

File: app/src/index.js [hunk 27]
```diff
@@ -2,11 +2,24 @@
...
+    // To avoid compile-time evaluation
+    eval("window.NODE_ENV=process.env.NODE_ENV; window.DIR_NAME = __dirname");
+    if (NODE_ENV == "development") {
+        static_path = "./"
+    } else {
+        static_path = DIR_NAME.replace(/app\.asar$/, 'static')+"/";
+    }
+} else {
+    static_path = "./"
+}
```
Comment (wdanilo) [inline app/src/index.js]: These ifs can be refactored to simpler cases, as 2 ifs have the same result.

File: app/src/main/index.js [hunk 28]
```diff
@@ -0,0 +1,73 @@
+'use strict'
+
+import { app, BrowserWindow } from 'electron'
+import * as path from 'path'
+import { format as formatUrl } from 'url'
```
Comment (wdanilo) [inline app/src/main/index.js]: alignment + sorting

File: app/src/main/index.js [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+
+const is_development = process.env.NODE_ENV !== 'production'
```
Comment (wdanilo) [inline app/src/main/index.js]: When the `NODE_ENV` is set to "production"? If I understand correctly, this is a runtime variable - whose responsibility is to set it? The user?

File: app/src/main/index.js [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+let main_window_keep_alive
+
+function create_main_window() {
+    const window = new BrowserWindow(
+        {
+            width          : 1024,
+            height         : 768,
+            frame          : false,
+        }
+    )
```
Comment (wdanilo) [inline app/src/main/index.js]: alignment

File: app/src/main/index.js [hunk 31]
```diff
@@ -0,0 +1,73 @@
...
+
+    var search = "";
```
Comment (wdanilo) [inline app/src/main/index.js]: never use `var`, use `let`

File: app/src/main/index.js (hunk 30 above)
Comment (wdanilo) [inline app/src/main/index.js]: Please use rust style here. We would write `{` in previous line.

File: app/src/main/index.js (hunk 31 above)
Comment (wdanilo) [inline app/src/main/index.js]: I would use better name here. As far as I see this is neither a search query not a search result. It is a scene name right?

File: app/src/main/index.js [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+    if (app.commandLine.hasSwitch("debug")) {
+        var value = app.commandLine.getSwitchValue("debug");
+        search = "?debug";
+        if (value !== undefined) {
```
Comment (wdanilo) [inline app/src/main/index.js]: if (value) { ...

File: app/src/main/index.js [hunk 33]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            width          : 1024,
+            height         : 768,
+            frame          : false,
+        }
+    )
+
+    var search = "";
+    if (app.commandLine.hasSwitch("debug")) {
+        var value = app.commandLine.getSwitchValue("debug");
```
Comment (wdanilo) [inline app/src/main/index.js]: You use literal `"debug"` 2 times in the code.

Comment (wdanilo) [inline app/src/main/index.js]: First you are checking if something exist, then yo uare getting it, then you are checking if its undefined. This is low quality code.

File: app/src/main/index.js [hunk 34]
```diff
@@ -0,0 +1,73 @@
...
+        search = "?debug";
+        if (value !== undefined) {
+            search += "=" + value
+        }
+    }
+
+    if (is_development) {
+        window.webContents.openDevTools();
+        var url = `http://localhost:${process.env.ELECTRON_WEBPACK_WDS_PORT}/main/` + search;
```
Comment (wdanilo) [inline app/src/main/index.js]: who is setting the `ELECTRON_WEBPACK_WDS_PORT` environment variable? Is it set automatically and can be overriden? Asking just to know it.

File: app/src/main/index.js [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+        window.loadURL(url)
+    } else {
+        window.loadURL(formatUrl({
+            pathname : path.join(__dirname.replace(/app\.asar$/, 'static'), '/main/index.html'),
+            protocol : 'file',
+            slashes  : true
+        }) + search)
```
Comment (wdanilo) [inline app/src/main/index.js]: This is a really complex expression. Please refactor it to variables. It would be so nicer to write `let url = formatUrl({pathname,protocol,slashes})` instead :)

File: app/src/main/index.js [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+    if (is_development) {
+        window.webContents.openDevTools();
+        var url = `http://localhost:${process.env.ELECTRON_WEBPACK_WDS_PORT}/main/` + search;
+        window.loadURL(url)
+    } else {
+        window.loadURL(formatUrl({
+            pathname : path.join(__dirname.replace(/app\.asar$/, 'static'), '/main/index.html'),
```
Comment (wdanilo) [inline app/src/main/index.js]: I see this code for the 2nd time.

File: app/src/main/index.js [hunk 37]
```diff
@@ -0,0 +1,73 @@
...
+    window.on('closed', () => {
+        main_window_keep_alive = null
+    })
+
+    window.webContents.on('devtools-opened', () => {
+        window.focus()
+        setImmediate(() => {
+            window.focus()
+        })
+    })
```
Comment (wdanilo) [inline app/src/main/index.js]: What does this code do? It is not obvious until you know electron internals. Should be refactored to nicely named function.

File: app/src/main/index.js [hunk 38]
```diff
@@ -0,0 +1,73 @@
...
+
+  return window
+}
+
+// quit application when all windows are closed
+app.on('window-all-closed', () => {
+    // on macOS it is common for applications to stay open until the user explicitly quits
```
Comment (wdanilo) [inline app/src/main/index.js]: Let's use real sentences in comments, please :)

File: app/src/main/index.js [hunk 39]
```diff
@@ -0,0 +1,73 @@
...
+        window.focus()
+        setImmediate(() => {
+            window.focus()
+        })
+    })
+
+  return window
+}
+
+// quit application when all windows are closed
```
Comment (wdanilo) [inline app/src/main/index.js]: I believe this comment is redundant. If you feel its necessary, lets refactor this to a function with proper name instead.

File: script/build.py [hunk 40]
```diff
@@ -0,0 +1,52 @@
...
+
+def write_file(path, contents):
+    with open(path, 'w') as f:
+        f.write(contents)
+
+
+def patch_file(path, patcher):
+    print(f'Patching {path}')
+    code_to_patch = read_file(path)
+    patched_code = patcher(code_to_patch)
```
Comment (wdanilo) [inline script/build.py]: alignment

File: script/build.py [hunk 41]
```diff
@@ -0,0 +1,52 @@
...
+shutil.move('target/web/gui_bg.wasm', 'target/web/gui.wasm')
+
+# TODO [mwu] It should be possible to drop gzip program dependency by using Python's gzip library.
+subprocess.check_call(['gzip', '--keep', '--best', '--force', 'target/web/gui.wasm'])
+
+# Note [MWU] We build to provisional location and patch files there before copying, so the backpack don't get errors
+#            from processing unpatched files.
+#            Also, here we copy into (overwriting), without removing old files. Backpack on Windows does not tolerate
+#            removing files it watches.
+copy_tree('target/web', 'app/src-rust-gen')
```
Comment (wdanilo) [inline script/build.py]: Please do not code in python on top-level. Move it to `main` function and evaluate it using python main magic instead :)

File: README.md [hunk 42]
```diff
@@ -36,27 +37,61 @@ In particular, you can provide them with `--release`, `--dev`, or `--profile`
...
-cd app
-npm install
-npm run start 
+ide/app$ npm install
+ide/app$ npm run web:dev 
 ```
 
 You can now navigate to http://localhost:8080 and play with it! The example
-scenes will be available at http://localhost:8080/debug.
+scenes will be available at http://localhost:8080/?debug.
```
Comment (wdanilo) [inline README.md]: Why this change? It is IMO much uglier than it was. I would like to have a final application in the cloud also following the convention that you can just go to `enso.org/app/debug` to access the debug screen. What are the rationales behind this change?

File: app/webpack.config.js (hunk 25 above)
Comment (notdanilo) [reply app/webpack.config.js]: I have tried to put the files to make them accessible in many different locations. Putting it in `src/renderer`, `src/common` or `src/main` doesn't make them easily accessible.

I found out `static` is a path defined by Electron to put static resources, such as binary files, image, etc... And it's fairly well documented. It was the way to go.

Putting the files directly in `static` wasn't a good option either. `index.html`, for instance, gets replaced by a generated `index.html` from Electron's renderer's `index.html`. So that's why I moved them to `static/main`.

File: app/webpack.config.js (hunk 26 above)
Comment (notdanilo) [reply app/webpack.config.js]: It's the content base to run the web development service, because all the files are moved to this path.

File: README.md (hunk 42 above)
Comment (wdanilo) [reply README.md]: I need to play with it. I will do it on evening.

File: app/src/main/index.js (hunk 29 above)
Comment (wdanilo) [reply app/src/main/index.js]: What are the possible options here? 'production' and what else? I would enumerate them in the code in order to keep track of possible options here. In case we will ever want to use it in more complex examples it would already be there. Plus, pls add comment that electron sets it.

File: app/src/main/index.js [hunk 43]
```diff
@@ -0,0 +1,67 @@
...
+            debug_scene += "=" + value
+        }
+    }
+
+    if (is_development) {
+        window.webContents.openDevTools();
+        let url = `http://localhost:${process.env.ELECTRON_WEBPACK_WDS_PORT}/main/` + debug_scene;
+        window.loadURL(url)
+    } else {
+        let static_path = __dirname.replace(/app\.asar$/, 'static');
```
Comment (wdanilo) [inline app/src/main/index.js]: I see this code in 2 places now. This is a little error prone. Plus I still do not understand what it does. Why you have to replace app.asar with static? What is the logic behind it? Could you please describe how electron uses this thing and why you need to replace it ?

File: app/src/main/index.js [hunk 44]
```diff
@@ -0,0 +1,67 @@
...
+        let url         = formatUrl({
+            pathname : path.join(static_path, '/main/index.html'),
```
Comment (wdanilo) [inline app/src/main/index.js]: This literal should be extracted to some global const

File: app/src/main/index.js [hunk 45]
```diff
@@ -0,0 +1,67 @@
...
+import * as path               from 'path'
+
+const is_development = process.env.NODE_ENV !== 'production'
+
+let main_window_keep_alive
+
+function create_main_window() {
+    const window = new BrowserWindow({
+        width  : 1024,
+        height : 768,
```
Comment (wdanilo) [inline app/src/main/index.js]: these literals should be extracted to some global consts

File: app/src/static_path.js [hunk 46]
```diff
@@ -0,0 +1,40 @@
...
+
+    // Main process
+    if (typeof process !== 'undefined' && typeof process.versions === 'object' && !!process.versions.electron) {
+        return true;
+    }
+
+    // Detect the user agent when the `nodeIntegration` option is set to true
+    if (typeof navigator === 'object' && typeof navigator.userAgent === 'string' && navigator.userAgent.indexOf('Electron') >= 0) {
+        return true;
+    }
```
Comment (wdanilo) [inline app/src/static_path.js]: these lines are too long plus there is a lot of repetition. Moreover, this seems like a really complex way of chekcing if we are in electron. Is there an official doc telling that this is the way to check it? If so, it should be linked in a comment.

File: app/src/static_path.js [hunk 47]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+/// Get static path.
+export function get_static_path(dirname) {
+//    if (is_electron() && !is_development()) {
+        return dirname.replace(/app\.asar$/, 'static');
+//    } else {
+//        return "./"
+//    }
+}
```
Comment (wdanilo) [inline app/src/static_path.js]: Here is some commented code. Again, I woudllike to know what is the logic behind repleacing app.asar ith static.

File: app/src/static_path.js [hunk 48]
```diff
@@ -0,0 +1,40 @@
...
+/// Check if the code is running in development mode.
+function is_development() {
+    if (is_electron()) {
+        return typeof process       !=  'undefined'
+            && typeof process.env   === 'object'
+            && process.env.NODE_ENV ==  'development';
+    } else {
+        return true;
+    }
+}
```
Comment (wdanilo) [inline app/src/static_path.js]: This documentation tells that this function checks if the code is running in debvelopment mode. But if we are not in electron it always returns true. This is strange. Why we cannot run in development mode in a browser? What is a development mode? It should be either fixed or docummented better..

File: app/webpack.config.js (hunk 26 above)
Comment (wdanilo) [reply app/webpack.config.js]: Danilo, please answer my question above. I'm asking about each line chagned. Could you please describe every line changed with information why it was changed like this? Why `-publicPath` why `-historyApiFallback` why `+ contentBase` and why `+hot` ?

File: script/build.py [hunk 49]
```diff
@@ -0,0 +1,60 @@
...
+    os.chdir(repo_root)
+
+    print('Building with wasm-pack...')
+    subprocess.check_call(['wasm-pack', 'build', '--target', 'web', '--no-typescript',
+                           '--out-dir', '../../target/web', 'lib/gui'])
+    patch_file('target/web/gui.js', js_workaround_patcher)
+    shutil.move('target/web/gui_bg.wasm', 'target/web/gui.wasm')
+
+    # TODO [mwu] It should be possible to drop gzip program dependency by using Python's gzip
+    #            library.
```
Comment (wdanilo) [inline script/build.py]: Are you sure the gzip lib in python has as good results as gzip shipped in the sys? I mean, is thisthe same impl?

File: script/build.py [hunk 50]
```diff
@@ -0,0 +1,60 @@
...
+    subprocess.check_call(['gzip', '--keep', '--best', '--force', 'target/web/gui.wasm'])
+
+    # Note [MWU] We build to provisional location and patch files there before copying, so the
+    #            backpack don't get errors from processing unpatched files.
+    #            Also, here we copy into (overwriting), without removing old files. Backpack on
+    #            Windows does not tolerate removing files it watches.
```
Comment (wdanilo) [inline script/build.py]: Regarding NOTE comments, I would not write [MWU] here. We write it next to todos in order to know who should clean up the code, but notes are just general information that should be kept in the code.

Moreover, keep NOTE in the first line and move the whole note to next line without indentation - I think thsi style is much better than keeping big useless indentaiton here.

File: script/build.py (hunk 49 above)
Comment (mwu-tow) [reply script/build.py]: I quite strongly expect so but I did not had time to research this, so I kept original logic intact and left note for future reconsideration.

File: script/build.py (hunk 50 above)
Comment (mwu-tow) [reply script/build.py]: done

File: script/build.py (hunk 49 above)
Comment (mwu-tow) [reply script/build.py]: i've left information in the comment to take care of that

File: README.md [hunk 51]
```diff
@@ -3,15 +3,11 @@
...
-# BaseGL
+# Enso Studio
 
-BaseGL is a blazing fast 2D vector rendering engine with a rich set of
-primitives and a GUI component library. It is able to display millions of shapes
-60 frames per second in a web browser on a modern laptop hardware. 
-
-This repository is a work in progress of BaseGL 2.0. Please refer to BaseGL 1.0
-repository for more information: https://github.com/luna/basegl-old.
+Enso Studio is an IDE for hybrid visual and textual functional programming.
```
Comment (mwu-tow) [inline README.md]: (I comment because I see this touched, but updating readme is not really part of this task)
There should be a mention that this is an IDE for Enso programming language, and that language engine development happens on the enso repository https://github.com/luna/enso.

File: README.md [hunk 52]
```diff
@@ -36,33 +32,67 @@ In particular, you can provide them with `--release`, `--dev`, or `--profile`
...
+on your system.
+
+### Web Application
+In order to build the IDE web application, follow the steps bellow:
 
 ```bash
-cd app
-npm install
-npm run start 
+ide/app$ npm install
```
Comment (mwu-tow) [inline README.md]: Don't we need to run the `build` script first to compile rust?

File: app/src/index.js [hunk 53]
```diff
@@ -2,8 +2,8 @@
 /// user with a visual representation of this process (welcome screen). It also implements a view
 /// allowing to choose a debug rendering test from.
 
-import * as loader_module from './loader'
-import * as html_utils    from './html_utils'
+import * as loader_module  from './loader'
+import * as html_utils     from './html_utils'
```
Comment (mwu-tow) [inline app/src/index.js]: Why additional spaces?

File: README.md [hunk 54]
```diff
@@ -36,33 +32,67 @@ In particular, you can provide them with `--release`, `--dev`, or `--profile`
...
+```
+
+If you want to distribute the desktop application, run the following steps instead:
+
+```bash
+ide/app$ npm install
+ide/app$ npm run electron:dist
+```
+
+The generated executable will be available at `dist/enso-ide-MAJOR.MINOR.*`. Its extension will
```
Comment (mwu-tow) [inline README.md]: I'd describe where are MAJOR.MINOR values defined.

File: README.md (hunk 16 above)
Comment (wdanilo) [reply README.md]: @notdanilo ?

File: README.md (hunk 52 above)
Comment (wdanilo) [reply README.md]: I believe that after the changes I requested above will be introduced (titles to the sections), this should be obvious.

File: README.md (hunk 42 above)
Comment (wdanilo) [reply README.md]: Do we still need it?

File: app/package.json [hunk 55]
```diff
@@ -1,17 +1,26 @@
...
     "create-wasm-app": ".bin/create-wasm-app.js"
   },
   "scripts": {
     "build": "webpack --mode production",
     "build-dev": "webpack --mode development",
-    "start": "webpack-dev-server --mode development",
+    "web:dev": "webpack-dev-server --mode development",
     "start-build-time-profile": "webpack-dev-server --mode development --profile --progress",
-    "prod-server": "npx http-server -p 3000 --proxy http://localhost:3000\\? dist"
+    "web:prod": "npx http-server -p 3000 --proxy http://localhost:3000\\? static/main",
```
Comment (wdanilo) [inline app/package.json]: As we have our own server impl now, this should either use our server impl or should be removed - using 2 server impls would bring confusion. For example, when downloading gziped content - it will work with one server but not with second one. I'm ok with removing it for now.

File: app/src/http-server.js [hunk 56]
```diff
@@ -0,0 +1,38 @@
...
+        let server = this;
+        fs.readFile(resource, function (err,data) {
+            if (err) {
+                server.process("/index.html", response);
+            } else {
+                let contentType   = mime.contentType(path.extname(resource))
+                let contentLength = data.length;
+                response.setHeader('Content-Type'  , contentType);
+                response.setHeader('Content-Length', contentLength);
+                response.writeHead(200);
```
Comment (wdanilo) [inline app/src/http-server.js]: What is this literal "200" here? Please refactor it to a variable.

File: app/src/main/index.js [hunk 57]
```diff
@@ -0,0 +1,93 @@
...
+
+// == Web Server ==
+
+const HTTP_SERVER_PORT = 9090;
+
+
+// === Execution mode ===
+
+const MODE_PRODUCTION  = 'production';
+const MODE_DEVELOPMENT = 'development';
```
Comment (wdanilo) [inline app/src/main/index.js]: just a minor request. Lets make it enum:
```javascript
const mode = {
    PRODUCTION  = 'production',
    DEVELOPMENT = 'development'
}
```

File: app/src/main/index.js [hunk 58]
```diff
@@ -0,0 +1,93 @@
...
+const IS_DEVELOPMENT   = process.env.NODE_ENV === MODE_DEVELOPMENT
+
+
+// === Window properties ===
+
+const WINDOW_WIDTH      = 1024;
+const WINDOW_HEIGHT     = 768;
+const WINDOW_FRAME      = false;
```
Comment (wdanilo) [inline app/src/main/index.js]: lets make it struct as well.

**Comment by wdanilo** [comment]
Will not be merged due to security issues and code quality issues.
